Add mbio eda phase 1 app helpers

asizemore opened the pull request:
_Resolves issue #6_

This PR will eventually add functions necessary for mbio phase 1 apps: alpha div, beta div, and relative abundance.

# To Dos
## Before finalizing implementation
- [x] Add jsd distance to beta div (may want to write this as an rcpp function? See [discussion](https://stackoverflow.com/questions/29050368/which-r-implementation-gives-the-fastest-jsd-matrix-computation)). 
- [x] If taxa is unknown, fill with "Unknown {taxonLevel-1}
- [x] Implement draft rel abundance
- [x] Draft 'writeToFile' type functions

## After finalizing
- [ ] Write tests.
- [ ] Write validation functions.
- [ ] Argument checks.
- [ ] More proactive error handling


## Updates
This PR is wip. Check back for updates.

**08.24.21** - still brainstorming. Have at least a basic idea of helpers.

**09.10.21** - basic functionality of beta and alpha div

**09.14.21** - relative abundance working, clean functions+scripts. Added ability to clean taxa names. Harder than expected.

asizemore committed:
placeholders for mbio app helpers

asizemore committed:
add makeDefaultTree

asizemore committed:
Merge branch 'main' into feature/6/add-mbio-apps

asizemore committed:
add utils

asizemore committed:
Merge branch 'main' into feature/6/add-mbio-apps

asizemore committed:
drafted working alpha div

asizemore commented:
Adding "scripts" to mbio folder.

asizemore committed:
add draft betadiv

asizemore committed:
Merge branch 'main' into feature/6/add-mbio-apps

asizemore committed:
Merge branch 'main' into feature/6/add-mbio-apps

asizemore committed:
rel abundance working

asizemore committed:
add rel abund working

asizemore committed:
reorganize and clean taxa names

d-callan commented on .gitignore:
```diff
@@ -1 +1,2 @@
-.DS_Store
\ No newline at end of file
+.DS_Store
+Bee_data/
```
? I dont actually see Bee data here. not sure thats the route we want to take for testing either..

d-callan commented on lib/R/Mbio/abundance.R:
```diff
@@ -0,0 +1,52 @@
+
+#   # This part seems redundant, but it's to get around using tidyr::complete (see shiny's otu_table.R line 18)
+#   otu <- makeOTU(df, taxonomicLevel)
+#   plot.data::logWithTime(paste("Created OTU table with", NROW(otu), "samples and", (NCOL(otu)-1), "taxa."), verbose)
```
hmmmmm.. this makes me think. maybe logWithTime should live somewhere else? not important right now i suppose. but if our list of generic helpers grow, we may want some sort of general purpose package for them.

asizemore replied:
+1 and let's add `matchArg` to that list of generic helpers

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,114 @@
+
+    # Clean data - rename, fill Nas with 0
+    #### Maybe this part should be a separate function of cleaning data?
+    data.table::setnames(df, c("Sample ID", "Relative Abundance", "Absolute Abundance", "Kingdom/SuperKingdom"), c("SampleID", "RelativeAbundance", "AbosluteAbundance", "Kingdom"), skip_absent = T)
```
what does the data you're testing w look like? i think from the merge service we can expect a column per taxa at some specified taxonomic level w rows per sample. so the columns would be things like 'taxa1', 'taxa2', etc and those taxa would all represent some genus say

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,114 @@
+    # Clean data - rename, fill Nas with 0
+    #### Maybe this part should be a separate function of cleaning data?
+    data.table::setnames(df, c("Sample ID", "Relative Abundance", "Absolute Abundance", "Kingdom/SuperKingdom"), c("SampleID", "RelativeAbundance", "AbosluteAbundance", "Kingdom"), skip_absent = T)
+    data.table::setnafill(df, fill = 0, cols = c("RelativeAbundance", "AbosluteAbundance"))
```
we need to think about this some. because doing it here means that only the compute apps impute NA to 0, and not the pass app.

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,114 @@
+    data.table::setnafill(df, fill = 0, cols = c("RelativeAbundance", "AbosluteAbundance"))
+    
+    # Replace column N/A with unknown taxonLevel -1. 
+    fullTaxonomy <- names(df)[3:(which(names(df)==taxonomicLevel))]
```
i think i remember shiny doing this, and get why its here. but i feel like we should talk w wojtek. this level of data cleaning should probably happen before it gets to the eda services

d-callan replied:
in fact i think it has to happen before this point no? idk how an NA value (or several) would look in the tree of vars for ex, or how we could reasonably request them as vars from the subset service.

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,114 @@
+
+    # Aggregate by taxonomic level
+    byCols <- c(taxonomicLevel, 'SampleID')
+    df <- df[, .("Abundance" = sum(RelativeAbundance)), by = eval(byCols)]
```
this is an interesting point. if we are operating at taxanomic level 'genus' and request a particular genus, what should we expect to get from the subset service? i think its either a) already a single value representing the sum of values across all species belonging to that genus or b) a multi-valued var w individual values for each species, which we would then have to sum like youre suggesting. i think we probably can and should go for option A, if we arent already (should check w wojtek). i dont think the subset influences this number right? its calculated per sample and can simply be loaded as a summed value?

asizemore replied:
Option A would be great, I didn't realize the subsetting service could do that! I'll assume this is how it works from now on.

d-callan replied:
I don't mean to suggest that I think the subset service does this summing for us. But it may be that wojtek has already summed them and loaded the sums. Should check w him how the data is structured in the db.

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,114 @@
+    return(topN)
+  }
+
+  makeOTU <- function(df, taxonomicLevel) {
```
we should be receiving otu tables from the subset/ merge service directly no?

d-callan commented on lib/R/VEuPathDB_General/utils.R:
```diff
@@ -0,0 +1,36 @@
+## General helper functions
+
+
+  # Write to file - will need to be a helper function
```
assuming this is for caching computations? we should check w steve ryan and ellie about where to expect to be able to write to/ how this will work etc.

d-callan commented on lib/R/VEuPathDB_General/utils.R:
```diff
@@ -0,0 +1,36 @@
+  # }
+
+
+  # Write to json ()
```
similar thought to writeDT.. think the idea is generally right that well need to do this, itll exist as a function somewhere, and the function will look roughly like this one. but being uncertain about details yet, and these being intended as general helpers i think id: 

1. remove references to computations/ metadata etc from the logging. make them as generic as possible
2. allow a directory/ path to be passed as an arg so we can store files wherever wed like to whatever name wed like. if nothing is passed, then maybe default to a tmp dir

d-callan replied:
i guess the other option is to make the function names much more specific. either way.. just so long as they dont have generic names and compute-specific intent.

d-callan commented on lib/R/VEuPathDB_General/utils.R:
```diff
@@ -0,0 +1,36 @@
+#     plot.data::logWithTime(paste('New metadata file written:', outFileName), verbose)
+#   }
+
+# Parsing args? Each app will have it's own list of args. Args come in via string or some other weirdness sometimes.
```
i think we should worry about validating each apps args in the compute service. then any mbio specific helpers can additionally validate their own args as part of the fxn you write. dont think (and certainly hope) we wont need a generic helper for it here

d-callan commented on lib/R/Mbio/abundance.R:
```diff
@@ -0,0 +1,52 @@
+#   rankingMethods <- c('median','max','q3','var')
+#   rankedTaxaList <- lapply(rankingMethods, topNTaxa, df=formattedDT, cutoff=10, taxonomicLevel = taxonomicLevel)
+
+#   # Admittedly terrible. Functions but needs clenaing up. Struggled to find a nice data.table solution
```
so assuming appending the ranking method to the column header is the method we ultimately use (which seems to be what youre doing here).. can topNTaxa return a data.table w those columns and the ranking method used already appended, rather than just a list of taxa?

d-callan replied:
it might even take the otu table format, rather than the reshaped thing, do whatever shaping is necessary to answer the question for itself, and then return the otu table format as well?

d-callan replied:
.. also possible ive misunderstood this code, not highly readable yet

asizemore replied:
re column names: i fought this for a while but came up with the above because I wanted to know the original column names as well since those would be used as the display names. I could also remove the appended info after appending it in the topNTaxa but that seemed not great either. 

Additionally i'm not sure appending the method is the best way to go about it anyhow. I also wrangled with the idea of returning the `rankedTaxaList` as computed variables, but that would require a more complicate api. However, it would be really nice to just return a list with one element per method that contains the top N taxa. Then, the output dt would only need to be the unique taxa that are in at least one of the top N lists. For example, in the bee data, it was something like only 15 unique taxa showed up in any of the four top N lists.

d-callan replied:
thats clever. i think id be fine w it. were already anticipating three endpoints for each compute, for the reason that these can be very variable from one compute to the next and also quite complicated. so im not opposed to a complicated api if that is also the best api. the dt i suppose would be returned with our data endpiont or whatever it ends up called, then the ranked taxa list list could be part of the metadata endpoint. and the app in the data service just needs to know about both of them, and how they relate.

d-callan commented on lib/R/Mbio/abundance.R:
```diff
@@ -0,0 +1,52 @@
+#   print(head(dt))
+#   print(jsonList)
+
+#   plot.data::logWithTime("Completed abundance app.", verbose)
```
should include any config here probably, like we did in plot.data

d-callan commented on lib/R/Mbio/beta_div.R:
```diff
@@ -0,0 +1,66 @@
+
+#   } else if (identical(method, 'jsd')) {
+
+#     # TO DO
```
i saw the link. think itd probably not be too great an effort to provide our own jsd function which was a hybrid of the two solutions i saw there. both properly vectorized, and leveraging existing rcpp fxns.

d-callan commented on lib/R/Mbio/beta_div.R:
```diff
@@ -0,0 +1,66 @@
+#   percentVar <- round(100*(eigenvecs / sum(eigenvecs)), 1)
+
+
+#   # Keep dims 1:20?
```
idk yet what well want in the future. but id start by just returning whatever we mean to use to plot w, and change it if we need later.

d-callan reviewed (commented):
very glad we did this, and think writing our own jsd fxn is reasonable and doable

d-callan commented:
also, on the general organization.. i think we want a function for each app, as opposed to scripts. could do similar to plot.data, where there is one fxn that puts together the dt to write, and one to wrap that fxn plus the write* helper(s).

asizemore committed:
start with otu

asizemore committed:
remove cleaning functions

asizemore committed:
wrap calculations into functions

asizemore committed:
reorganized outputs

asizemore committed:
Merge branch 'main' into feature/6/add-mbio-apps

d-callan commented on lib/R/Mbio/abundance.R:
```diff
@@ -0,0 +1,50 @@
+rankedAbundance <- function(otu, method = c('median','max','q3','var'), verbose = c(TRUE, FALSE)) {
+
+    # Initialize and check inputs
+    verbose <- plot.data::matchArg(verbose)
```
validate method arg?

d-callan commented on lib/R/Mbio/mbio-utils.R:
```diff
@@ -0,0 +1,64 @@
+topNTaxa <- function(df, method=c('median','max','q3','var'), cutoff=10, taxonomicLevel="TaxonomicLevel") {
+    
+    ## Rank by method
+    #### Eventually change TaxonomicLevel -- decide if we'll ever need to actually pass this info or not.
```
i think probably we dont need taxonomic level. im imagining that by this point the only taxa we have in hand are for the level requested.

d-callan commented on lib/R/Mbio/abundance.R:
```diff
@@ -0,0 +1,50 @@
+
+    ## Ignoring anything past here for now...
+
+    # placeholder
```
we should probably talk w steve about our idea of caching multiple ranking methods, see what he thinks of that/ how he wants to handle it.

d-callan commented on lib/R/Mbio/alpha_div.R:
```diff
@@ -0,0 +1,65 @@
+    data.table::setnames(dt, 'alphaDiv', method)
+
+    results <- list(
+      'dt' = dt,
```
you might consider trying to standardize the format of this results list. maybe a 'data', 'stats', and 'metadata' entry. and those can each be a data.table or list themselves as necessary. then we could consider having a single write* helper that knows to look for each of those entries in the results list. (this is approaching oo territory, but we can punt on that for now)

asizemore committed:
clean and reorganize ranktaxa

asizemore committed:
reorganize outupt list

asizemore commented:
note so i don't forget -- needed to allow 10GB memory for the container in order to not crash when running jsd...

asizemore committed:
functioning jsd

asizemore committed:
attempt to improve jsd

asizemore committed:
working jsd

asizemore committed:
add general write json function

asizemore commented:
Right now we can get away with having a general write to json function since we're just writing basic attributes and such. I'm also unboxing properties when they're added to the list instead of during writing which is probably cheating.

asizemore committed:
add draft api for reference

asizemore commented:
for fun i added a draft of the api  -- i'll remove it before merging of course but might be helpful for us both to be able to reference and change

asizemore commented:
@d-callan ready for a new look when you have a min!

d-callan commented on Dockerfile:
```diff
@@ -23,6 +23,8 @@ RUN R -e "install.packages('phytools')"
 RUN R -e "install.packages('remotes')"
 RUN R -e "install.packages('vegan')"
+RUN R -e "install.packages('Rcpp')"
+RUN R -e "install.packages('stringr')"
```
consider stringi instead? i remember it being generally faster and having more features..

d-callan commented on cs-api-draft.raml:
```diff
@@ -0,0 +1,57 @@
+    properties:
+      wasCutoff:
+        type: string
+        enum: ['TRUE', 'FALSE']
```
so i have also used the enum TRUE/FALSE , particularly when i know the inputs are going to eventually land in R. but i think there may be a boolean type w true/false values. dont really remember for sure, but worth looking into if this is going to js instead..

d-callan commented on cs-api-draft.raml:
```diff
@@ -0,0 +1,57 @@
+      data:
+        type: alphaDivData # some data table with samples as rows
+      appMetadata: 
+        type: metadata[]
```
should probably be called something other than simply metadata

d-callan commented on cs-api-draft.raml:
```diff
@@ -0,0 +1,57 @@
+      computedAxisLabel: 
+        type: string   # Ex. ‘Alpha Diversity’
+        required: false
+      defaultRange:
```
seems this is meant to be pretty general purpose. is it safe to assume all computed vars are continuous? maybe fine for now anyhow, and we can always come back to it if needed.

d-callan commented on cs-api-draft.raml:
```diff
@@ -0,0 +1,57 @@
+    properties:
+      computedVariables:
+        type: string[]    # Ex. ['shannon’]
+      computedVariableLabels:
```
do we need a computed var label and axis label? seems confusing, conflating the ideas of an app and a viz maybe. could there just be one label, something like 'Alpha Diversity: Shannon'

asizemore replied:
In alpha div scatter, that's true they'd be the same. But for abundance box, the computedVarLabels would be the taxa names and the axis label is Relative Abundance. Similarly, if we let alpha div vars comprise a collection, they'd need to retain their computedVarLabels for box tick labels, for example, and then the axis title could be "Alpha Diversity".
For now, abundance is the outlier, so we could add this info elsewhere or at the very least make computedAxisLabel optional.

d-callan commented on cs-api-draft.raml:
```diff
@@ -0,0 +1,57 @@
+        required: false
+      computeDetails:
+        type: string      # Ex. 'No convergence after 100 iterations' or ‘134 of 150 samples have sufficient data for this computation’
+      computeName:
```
should maybe split this into further objects, as you have metadata at various conceptual levels here.. perhaps a ComputedVariableMetadata object and a parent ComputeAppMetadata object?

d-callan commented on lib/R/Mbio/alpha_div.R:
```diff
@@ -0,0 +1,78 @@
+
+    # Merge all data into one data table and remove from computeResults
+    dtList <- lapply(computeResults, function(x) return(x$dt))
+    dt <- Reduce(function(...) merge(..., all = TRUE, by='SampleID'), dtList)
```
not sure i like this. feels like we may want a dt w attributes again?

d-callan replied:
these return objects are becoming more complex than i anticipated. we should pause and develop a more formal strategy/ structure i think.

asizemore replied:
i like dt with attributes for sure

d-callan commented on lib/preload.R:
```diff
@@ -4,8 +4,14 @@ library(vegan)
 invisible(sapply(files.sources, source))
+
+# Compiled functions
+comp.files.sources <- list.files('opt/rserve/lib/src/', full.names=TRUE, recursive=TRUE)
```
i guess im fine leaving this feature in rserve, despite the fact that its need is making me lean towards a package for the mbio features.

d-callan commented on lib/R/Mbio/pcoa.R:
```diff
@@ -0,0 +1,182 @@
+# From https://github.com/emmanuelparadis/ape/blob/master/R/pcoa.R on 10.13.21
+# This latest version addresses a pcoa rounding issue that we experienced: https://github.com/emmanuelparadis/ape/issues/17
+PCOA <- function(D, correction="none", rn=NULL)
```
why do we need our own pcoa? i think i missed something..

d-callan reviewed (commented):
seeing us add more dependencies, writing our own distance functions, increasing the complexity of the resulting object etc is making me feel like we may want to make this an R package sooner rather than later. i really hoped to avoid it for phase 1 at least, but testing and performance and organization are becoming increasingly important. 

lets decide what wed like to call it, ill work on getting a skeletal veupath repo for it. then i think you should flesh that out as much as possible while i return to the data service.

d-callan commented:
https://github.com/VEuPathDB/microbiomeComputations
is now a thing. we should move most of this logic there. the only thing i think should stay in this pr are generic helpers used to format the results in such a way as to meet the compute service api

d-callan commented:
And if those end up not existing for whatever reason then we can close w/o merging. Sorry.

d-callan committed:
move compute service draft api to data service

this is likely to also be a short-term workaround until we have a functional compute service. but the data service is where the work is going to happen in the mean time

asizemore committed:
removed mbcomputr functions

d-callan reviewed (commented):
i love how clean this just got. but i think we should wait until we have the mbio R package at least building before merging.

asizemore commented:
Perfect sounds great!

asizemore committed:
add veupathUtils and clean

d-callan commented on Dockerfile:
```diff
@@ -23,13 +23,17 @@ RUN R -e "install.packages('phytools')"
 RUN R -e "install.packages('rbiom')"
 RUN R -e "install.packages('remotes')"
 RUN R -e "install.packages('vegan')"
```
can we remove vegan, rbiom etc from here?

asizemore replied:
yup! Will do!

asizemore replied:
i'll also add a microbiomeComputations version once the open PR there is merged

d-callan committed:
update installed packages

d-callan committed:
update attached libraries

d-callan committed:
Merge branch 'main' into feature/6/add-mbio-apps

d-callan reviewed (approved):
remember to give microbiomeComputations a version is the only thing.

asizemore committed:
microbiomeComputations v1.0.0

asizemore commented:
Added microbiomeComputations v1.0.0 and tested all apps with default parameters (includes all methods). Looking good!
